librz/analysis/arch/arm/arm_il64.c

```diff
@@ -0,0 +1,2713 @@
+		is_signed = true;
+		// fallthrough
+	case ARM64_EXT_UXTX:
+		src_bits = 64;
+		break;
+
+	default:
+		if (dst_bits == v_bits) {
+			return v;
+		} else {
+			return adjust_unsigned(dst_bits, v);
+		}
+	}
+
+	v = adjust_unsigned(src_bits, v);
+	return is_signed ? SIGNED(dst_bits, v) : UNSIGNED(dst_bits, v);
+}
+
+static RzILOpBitVector *apply_shift(arm64_shifter sft, ut32 dist, RZ_OWN RzILOpBitVector *v) {
+	if (!dist) {
+		return v;
+	}
+	switch (sft) {
+	case ARM64_SFT_LSL:
+		return SHIFTL0(v, UN(6, dist));
+	case ARM64_SFT_LSR:
+		return SHIFTR0(v, UN(6, dist));
+	case ARM64_SFT_ASR:
+		return SHIFTRA(v, UN(6, dist));
+	default:
+		return v;
+	}
+}
+
+#define REG(n)       read_reg(REGID(n))
+#define REGBITS(n)   reg_bits(REGID(n))
+#define MEMBASEID(x) insn->detail->arm64.operands[x].mem.base
+#define MEMBASE(x)   read_reg(MEMBASEID(x))
+
+/**
+ * IL to write a value to the given capstone reg
+ */
+static RzILOpEffect *write_reg(arm64_reg reg, RZ_OWN RZ_NONNULL RzILOpBitVector *v) {
+	rz_return_val_if_fail(v, NULL);
+	const char *var = reg_var_name(reg);
+	if (!var) {
+		rz_il_op_pure_free(v);
+		return NULL;
+	}
+	if (is_wreg(reg)) {
+		v = UNSIGNED(64, v);
+	}
+	return SETG(var, v);
+}
+
+static RzILOpBitVector *arg_mem(RzILOpBitVector *base_plus_disp, cs_arm64_op *op) {
+	if (op->mem.index == ARM64_REG_INVALID) {
+		return base_plus_disp;
+	}
+	RzILOpBitVector *index = read_reg(op->mem.index);
+	index = extend(64, op->ext, index, reg_bits(op->mem.index));
+	index = apply_shift(op->shift.type, op->shift.value, index);
+	return ADD(base_plus_disp, index);
+}
+
+/**
+ * IL to retrieve the value of the \p n -th arg of \p insn
+ * \p bits_inout Setting the backing variable to non-0 indicates that the result must have this bitness.
+ *               This is necessary for immediate operands for example.
+ *               In any case, if a value is returned, its bitness is written back into this storage.
+ */
+static RzILOpBitVector *arg(cs_insn *insn, size_t n, ut32 *bits_inout) {
+	ut32 bits_requested = bits_inout ? *bits_inout : 0;
+	cs_arm64_op *op = &insn->detail->arm64.operands[n];
+	switch (op->type) {
+	case ARM64_OP_REG: {
+		if (!bits_requested) {
+			bits_requested = REGBITS(n);
+			if (bits_inout) {
+				*bits_inout = bits_requested;
+			}
+		}
+		RzILOpBitVector *r = REG(n);
+		if (!r) {
+			return NULL;
+		}
+		return apply_shift(op->shift.type, op->shift.value, extend(bits_requested, op->ext, r, REGBITS(n)));
+	}
+	case ARM64_OP_IMM: {
+		if (!bits_requested) {
+			return NULL;
+		}
+		ut64 val = IMM(n);
+		if (op->shift.type == ARM64_SFT_LSL) {
+			val <<= op->shift.value;
+		}
+		return UN(bits_requested, val);
+	}
+	case ARM64_OP_MEM: {
+		RzILOpBitVector *addr = MEMBASE(n);
+		st64 disp = MEMDISP(n);
+		if (disp > 0) {
+			addr = ADD(addr, U64(disp));
+		} else if (disp < 0) {
+			addr = SUB(addr, U64(-disp));
+		}
+		return arg_mem(addr, &insn->detail->arm64.operands[n]);
+	}
+	default:
+		break;
+	}
+	return NULL;
```

Comment:
warn_if_reached()

Reply:
No, `NULL` is valid here for anything we don't handle yet. It is checked by callers.